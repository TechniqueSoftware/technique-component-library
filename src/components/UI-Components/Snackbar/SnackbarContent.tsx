import * as React from 'react';
import clsx from 'clsx';

import MuiSnackbarContent from '@material-ui/core/SnackbarContent';

import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';

import green from '@material-ui/core/colors/lightGreen';
import yellow from '@material-ui/core/colors/yellow';
import red from '@material-ui/core/colors/red';
import IconButton from '@material-ui/core/IconButton';
import WarningIcon from '@material-ui/icons/Warning';
import SuccessIcon from '@material-ui/icons/CheckCircle';
import ClubOSErrorIcon from '@material-ui/icons/Error';
import ClubOSInfoIcon from '@material-ui/icons/Info';
import CloseIcon from '@material-ui/icons/Close';

import { Theme, makeStyles, withStyles } from '@material-ui/core/styles';

export interface SnackbarContentProps {
  id: string;
  className?: string;
  'aria-describe-id'?: string;  // Default: 'snackbar'
  variant: 'success' | 'warning' | 'error' | 'info';
  message: React.ReactNode;
  maxWidth?: number;
  onUndo?: () => void;
  onClose?: () => void;
}

const useStyles = makeStyles((theme: Theme) => {
  return ({
    success: {},
    error: {},
    info: {},
    warning: {},
    icon: {
      fontSize: 20,
    },
    iconVariant: {
      opacity: 0.9,
    },
    undoLink: {
      fontWeight: 900,
    },
    message: {
      display: 'flex',
      alignItems: 'center',
      fontWeight: 600,
    },
  });
});

const clubOSWarningIconStyles = (theme: Theme) => ({
  root: {
    color: yellow[800],
    marginRight: theme.spacing(1),
    fontSize: 20,
  },
});

const clubOSSuccessIconStyles = (theme: Theme) => ({
  root: {
    color: green[600],
    marginRight: theme.spacing(1),
    fontSize: 20,
  },
});
const clubOSInfoIconStyles = (theme: Theme) => ({
  root: {
    color: theme.palette.primary.main,
    marginRight: theme.spacing(1),
    fontSize: 20,
  },
});

const clubOSErrorIconStyles = (theme: Theme) => ({
  root: {
    color: red[600],
    marginRight: theme.spacing(1),
    fontSize: 20,
  },
});

const ClubOSWarningIcon = (props: { classes: { root: any } }) => <WarningIcon className={props.classes.root} />;
const ClubOSSuccessIcon = (props: { classes: { root: any } }) => <SuccessIcon className={props.classes.root} />;

const ClubOSWarningIconStyles = withStyles(clubOSWarningIconStyles)(ClubOSWarningIcon);
const ClubOSSuccessIconStyles = withStyles(clubOSSuccessIconStyles)(ClubOSSuccessIcon);
const ClubOSErrorIconStyles = withStyles(clubOSErrorIconStyles)(ClubOSErrorIcon);
const ClubOSInfoIconStyles = withStyles(clubOSInfoIconStyles)(ClubOSInfoIcon);

const variantIcon = {
  success: ClubOSSuccessIconStyles,
  warning: ClubOSWarningIconStyles,
  info: ClubOSInfoIconStyles,
  error: ClubOSErrorIconStyles,
};

export interface SnackbarActionsProps {
  onUndo?: () => void;
  onClose?: () => void;
}

export function SnackbarAction(props: SnackbarActionsProps) {

  const { onUndo, onClose } = props;
  const classes = useStyles();
  return <React.Fragment>
    {!!onUndo &&
    <Button
      color="primary"
      size="small"
      onClick={onUndo}
    >
      Undo
    </Button>}
    {!!onClose &&
    <IconButton
      key="close"
      aria-label="Close"
      color="inherit"
      onClick={onClose}
    >
      <CloseIcon className={classes.icon} />
    </IconButton>
    }
  </React.Fragment>;
}

export const SnackbarContent = React.forwardRef((props: SnackbarContentProps, ref) => {
  const classes = useStyles();
  const {
    id,
    className,
    'aria-describe-id': ariaDescribeId = 'snackbar',
    message,
    maxWidth = 400,
    variant,
    onClose,
    onUndo
  } = props;

  const Icon = variantIcon[variant];
  const applyAriaDescribeId = ariaDescribeId || 'snackbar-content';

  return (
    <div ref={ref as any} id={id}>
      <Box maxWidth={maxWidth}>
        <MuiSnackbarContent
          className={clsx(classes[variant], className)}
          aria-describedby={applyAriaDescribeId}
          message={
            <span id={applyAriaDescribeId} className={classes.message}>
              <Icon className={clsx(classes.icon, classes.iconVariant)} />
              {message}
            </span>
          }
          action={<SnackbarAction onClose={onClose} onUndo={onUndo} />}
        />
      </Box>
    </div>
  );
});
