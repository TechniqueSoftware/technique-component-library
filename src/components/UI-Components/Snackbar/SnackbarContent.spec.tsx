import { SnackbarContent, SnackbarContentProps } from './SnackbarContent';
import { mount } from 'enzyme';
import * as React from 'react';

describe('SnackbarContent', () => {
  test('should render with minimal', () => {
    const props: SnackbarContentProps = {
      id: 'mockId',
      message: 'Mock message',
      variant: 'info',
      'aria-describe-id': 'id-9247924',
      onUndo: () => undefined,
    };
    const shallowWrapper = mount(
      <SnackbarContent {...props}/>
    );
    expect(shallowWrapper.debug()).toMatchSnapshot();
  });

  test('should render with all props', () => {
    const props: SnackbarContentProps = {
      id: 'mockId',
      'aria-describe-id': 'id-9247924',
      variant: 'info',
      maxWidth: 500,
      message: 'Mock Archived',
      onUndo: () => undefined,
      onClose: () => undefined,
    };

    const shallowWrapper = mount(
      <SnackbarContent {...props}/>
    );
    expect(shallowWrapper.debug()).toMatchSnapshot();
  });
});
