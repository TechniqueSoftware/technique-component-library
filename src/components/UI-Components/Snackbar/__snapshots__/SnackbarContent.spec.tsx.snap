// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnackbarContent should render with all props 1`] = `
"<ForwardRef id=\\"mockId\\" aria-describe-id=\\"id-9247924\\" variant=\\"info\\" maxWidth={500} message=\\"Mock Archived\\" onUndo={[Function: onUndo]} onClose={[Function: onClose]}>
  <div id=\\"mockId\\">
    <Styled(MuiBox) maxWidth={500}>
      <div className=\\"MuiBox-root MuiBox-root-3\\">
        <WithStyles(ForwardRef(SnackbarContent)) className=\\"\\" aria-describedby=\\"id-9247924\\" message={{...}} action={{...}}>
          <ForwardRef(SnackbarContent) classes={{...}} className=\\"\\" aria-describedby=\\"id-9247924\\" message={{...}} action={{...}}>
            <WithStyles(ForwardRef(Paper)) role=\\"alert\\" square={true} elevation={6} className=\\"MuiSnackbarContent-root\\" aria-describedby=\\"id-9247924\\">
              <ForwardRef(Paper) classes={{...}} role=\\"alert\\" square={true} elevation={6} className=\\"MuiSnackbarContent-root\\" aria-describedby=\\"id-9247924\\">
                <div className=\\"MuiPaper-root MuiSnackbarContent-root MuiPaper-elevation6\\" role=\\"alert\\" aria-describedby=\\"id-9247924\\">
                  <div className=\\"MuiSnackbarContent-message\\">
                    <span id=\\"id-9247924\\" className={{...}}>
                      <WithStyles(memo) className=\\"fontSize opacity\\">
                        <ForwardRef(InfoIcon) classes={{...}} className=\\"fontSize opacity\\">
                          <WithStyles(ForwardRef(SvgIcon)) classes={{...}} className=\\"fontSize opacity\\">
                            <ForwardRef(SvgIcon) classes={{...}} className=\\"fontSize opacity\\">
                              <svg className=\\"MuiSvgIcon-root memo-root-2 fontSize opacity\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" color={[undefined]} aria-hidden={true} role={[undefined]}>
                                <path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z\\" />
                              </svg>
                            </ForwardRef(SvgIcon)>
                          </WithStyles(ForwardRef(SvgIcon))>
                        </ForwardRef(InfoIcon)>
                      </WithStyles(memo)>
                      Mock Archived
                    </span>
                  </div>
                  <div className=\\"MuiSnackbarContent-action\\">
                    <SnackbarAction onClose={[Function: onClose]} onUndo={[Function: onUndo]}>
                      <WithStyles(ForwardRef(Button)) color=\\"primary\\" size=\\"small\\" onClick={[Function: onUndo]}>
                        <ForwardRef(Button) classes={{...}} color=\\"primary\\" size=\\"small\\" onClick={[Function: onUndo]}>
                          <WithStyles(ForwardRef(ButtonBase)) className=\\"MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall\\" component=\\"button\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"Mui-focusVisible\\" type=\\"button\\" onClick={[Function: onUndo]}>
                            <ForwardRef(ButtonBase) classes={{...}} className=\\"MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall\\" component=\\"button\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"Mui-focusVisible\\" type=\\"button\\" onClick={[Function: onUndo]}>
                              <button className=\\"MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall\\" onBlur={[Function (anonymous)]} onClick={[Function: onUndo]} onFocus={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} onKeyUp={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseUp={[Function (anonymous)]} onDragLeave={[Function (anonymous)]} onTouchEnd={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} tabIndex={0} type=\\"button\\" disabled={false}>
                                <span className=\\"MuiButton-label\\">
                                  Undo
                                </span>
                                <WithStyles(memo) center={false}>
                                  <ForwardRef(TouchRipple) classes={{...}} center={false}>
                                    <span className=\\"MuiTouchRipple-root\\">
                                      <TransitionGroup component={{...}} exit={true} childFactory={[Function: childFactory]} />
                                    </span>
                                  </ForwardRef(TouchRipple)>
                                </WithStyles(memo)>
                              </button>
                            </ForwardRef(ButtonBase)>
                          </WithStyles(ForwardRef(ButtonBase))>
                        </ForwardRef(Button)>
                      </WithStyles(ForwardRef(Button))>
                      <WithStyles(ForwardRef(IconButton)) aria-label=\\"Close\\" color=\\"inherit\\" onClick={[Function: onClose]}>
                        <ForwardRef(IconButton) classes={{...}} aria-label=\\"Close\\" color=\\"inherit\\" onClick={[Function: onClose]}>
                          <WithStyles(ForwardRef(ButtonBase)) className=\\"MuiIconButton-root MuiIconButton-colorInherit\\" centerRipple={true} focusRipple={true} disabled={false} aria-label=\\"Close\\" onClick={[Function: onClose]}>
                            <ForwardRef(ButtonBase) classes={{...}} className=\\"MuiIconButton-root MuiIconButton-colorInherit\\" centerRipple={true} focusRipple={true} disabled={false} aria-label=\\"Close\\" onClick={[Function: onClose]}>
                              <button className=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit\\" onBlur={[Function (anonymous)]} onClick={[Function: onClose]} onFocus={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} onKeyUp={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseUp={[Function (anonymous)]} onDragLeave={[Function (anonymous)]} onTouchEnd={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} tabIndex={0} type=\\"button\\" disabled={false} aria-label=\\"Close\\">
                                <span className=\\"MuiIconButton-label\\">
                                  <ForwardRef(CloseIcon) className={{...}}>
                                    <WithStyles(ForwardRef(SvgIcon)) className={{...}}>
                                      <ForwardRef(SvgIcon) classes={{...}} className={{...}}>
                                        <svg className=\\"MuiSvgIcon-root fontSize\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" color={[undefined]} aria-hidden={true} role={[undefined]}>
                                          <path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\" />
                                        </svg>
                                      </ForwardRef(SvgIcon)>
                                    </WithStyles(ForwardRef(SvgIcon))>
                                  </ForwardRef(CloseIcon)>
                                </span>
                                <WithStyles(memo) center={true}>
                                  <ForwardRef(TouchRipple) classes={{...}} center={true}>
                                    <span className=\\"MuiTouchRipple-root\\">
                                      <TransitionGroup component={{...}} exit={true} childFactory={[Function: childFactory]} />
                                    </span>
                                  </ForwardRef(TouchRipple)>
                                </WithStyles(memo)>
                              </button>
                            </ForwardRef(ButtonBase)>
                          </WithStyles(ForwardRef(ButtonBase))>
                        </ForwardRef(IconButton)>
                      </WithStyles(ForwardRef(IconButton))>
                    </SnackbarAction>
                  </div>
                </div>
              </ForwardRef(Paper)>
            </WithStyles(ForwardRef(Paper))>
          </ForwardRef(SnackbarContent)>
        </WithStyles(ForwardRef(SnackbarContent))>
      </div>
    </Styled(MuiBox)>
  </div>
</ForwardRef>"
`;

exports[`SnackbarContent should render with minimal 1`] = `
"<ForwardRef id=\\"mockId\\" message=\\"Mock message\\" variant=\\"info\\" aria-describe-id=\\"id-9247924\\" onUndo={[Function: onUndo]}>
  <div id=\\"mockId\\">
    <Styled(MuiBox) maxWidth={400}>
      <div className=\\"MuiBox-root MuiBox-root-1\\">
        <WithStyles(ForwardRef(SnackbarContent)) className=\\"\\" aria-describedby=\\"id-9247924\\" message={{...}} action={{...}}>
          <ForwardRef(SnackbarContent) classes={{...}} className=\\"\\" aria-describedby=\\"id-9247924\\" message={{...}} action={{...}}>
            <WithStyles(ForwardRef(Paper)) role=\\"alert\\" square={true} elevation={6} className=\\"MuiSnackbarContent-root\\" aria-describedby=\\"id-9247924\\">
              <ForwardRef(Paper) classes={{...}} role=\\"alert\\" square={true} elevation={6} className=\\"MuiSnackbarContent-root\\" aria-describedby=\\"id-9247924\\">
                <div className=\\"MuiPaper-root MuiSnackbarContent-root MuiPaper-elevation6\\" role=\\"alert\\" aria-describedby=\\"id-9247924\\">
                  <div className=\\"MuiSnackbarContent-message\\">
                    <span id=\\"id-9247924\\" className={{...}}>
                      <WithStyles(memo) className=\\"fontSize opacity\\">
                        <ForwardRef(InfoIcon) classes={{...}} className=\\"fontSize opacity\\">
                          <WithStyles(ForwardRef(SvgIcon)) classes={{...}} className=\\"fontSize opacity\\">
                            <ForwardRef(SvgIcon) classes={{...}} className=\\"fontSize opacity\\">
                              <svg className=\\"MuiSvgIcon-root memo-root-2 fontSize opacity\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" color={[undefined]} aria-hidden={true} role={[undefined]}>
                                <path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z\\" />
                              </svg>
                            </ForwardRef(SvgIcon)>
                          </WithStyles(ForwardRef(SvgIcon))>
                        </ForwardRef(InfoIcon)>
                      </WithStyles(memo)>
                      Mock message
                    </span>
                  </div>
                  <div className=\\"MuiSnackbarContent-action\\">
                    <SnackbarAction onClose={[undefined]} onUndo={[Function: onUndo]}>
                      <WithStyles(ForwardRef(Button)) color=\\"primary\\" size=\\"small\\" onClick={[Function: onUndo]}>
                        <ForwardRef(Button) classes={{...}} color=\\"primary\\" size=\\"small\\" onClick={[Function: onUndo]}>
                          <WithStyles(ForwardRef(ButtonBase)) className=\\"MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall\\" component=\\"button\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"Mui-focusVisible\\" type=\\"button\\" onClick={[Function: onUndo]}>
                            <ForwardRef(ButtonBase) classes={{...}} className=\\"MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall\\" component=\\"button\\" disabled={false} focusRipple={true} focusVisibleClassName=\\"Mui-focusVisible\\" type=\\"button\\" onClick={[Function: onUndo]}>
                              <button className=\\"MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall\\" onBlur={[Function (anonymous)]} onClick={[Function: onUndo]} onFocus={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} onKeyUp={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseUp={[Function (anonymous)]} onDragLeave={[Function (anonymous)]} onTouchEnd={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} tabIndex={0} type=\\"button\\" disabled={false}>
                                <span className=\\"MuiButton-label\\">
                                  Undo
                                </span>
                                <WithStyles(memo) center={false}>
                                  <ForwardRef(TouchRipple) classes={{...}} center={false}>
                                    <span className=\\"MuiTouchRipple-root\\">
                                      <TransitionGroup component={{...}} exit={true} childFactory={[Function: childFactory]} />
                                    </span>
                                  </ForwardRef(TouchRipple)>
                                </WithStyles(memo)>
                              </button>
                            </ForwardRef(ButtonBase)>
                          </WithStyles(ForwardRef(ButtonBase))>
                        </ForwardRef(Button)>
                      </WithStyles(ForwardRef(Button))>
                    </SnackbarAction>
                  </div>
                </div>
              </ForwardRef(Paper)>
            </WithStyles(ForwardRef(Paper))>
          </ForwardRef(SnackbarContent)>
        </WithStyles(ForwardRef(SnackbarContent))>
      </div>
    </Styled(MuiBox)>
  </div>
</ForwardRef>"
`;
